s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala

```diff
@@ -0,0 +1,98 @@
+package com.bluelabs.s3stream
+
+import java.io.FileOutputStream
+import java.io.RandomAccessFile
+import java.nio.channels.FileChannel
+import java.nio.file.Files
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.NotUsed
+import akka.stream.ActorAttributes
+import akka.stream.Attributes
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.scaladsl.FileIO
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.util.ByteString
+import java.nio.file.Path
+
+/**
+ * Buffers the complete incoming stream into a file, which can then be read several times afterwards.
+ * 
+ * The stage waits for the incoming stream to complete. After that, it emits a single Chunk item on its output. The Chunk
+ * contains a bytestream source that can be materialized multiple times, and the total size of the file.
+ *  
+ * @param maxMaterializations Number of expected materializations for the completed chunk. After this, the temp file is deleted.
+ * @param maximumSize Maximum size on disk to buffer
+ */
+class DiskBuffer (maxMaterializations: Int, maxSize: Int, tempPath: Option[Path]) extends GraphStage[FlowShape[ByteString, Chunk]] {
```

Comment:
Yeah -- I'm a bit torn on it for the same reason. It's nice having the clean "sign and issue an AWS request" bit relatively contained. Maybe hashing on the way out would make sense, but doing it with one read by making the request-generation take a Source[Bytestring] that we can read-and-hash at the same time.